src/Workspaces/Core/Portable/Execution/Serialization/Checksum.cs

```diff
@@ -0,0 +1,115 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Execution
+{
+    /// <summary>
+    /// Checksum of data
+    /// </summary>
+    internal sealed partial class Checksum : IObjectWritable, IEquatable<Checksum>
+    {
+        private readonly ImmutableArray<byte> _checkSum;
+        private int? _lazyHash;
+
+        public Checksum(byte[] checksum) :
+            this(ImmutableArray.Create(checksum))
+        {
+        }
+
+        public Checksum(ImmutableArray<byte> checksum)
+        {
+            _lazyHash = null;
+            _checkSum = checksum;
+        }
+
+        public byte[] ToArray()
+        {
+            // TODO: think a way to make this better
+            return _checkSum.ToArray();
```

Comment:
we will see whether this will cause too many allocation or not. right now, I dont believe this is the top offender. which can clean it up once it shows up as top offender.